File: _inc/laravel/app/Http/Controllers/Companies/BankAccountController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{BanksConstants, DatabaseConstants, UsersConstants};
use App\Models\{
    BankAccount,
    BillPayment,
    ChartOfAccount,
    CustomField,
    InvoicePayment,
    Payment,
    Revenue,
    Transaction
};
use App\Traits\ChecksLogin;
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Contracts\View\View;
use Illuminate\Http\{Request, RedirectResponse};
use Illuminate\Support\Facades\{Auth, DB, Log, Validator, View as ViewFacade};

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
class BankAccountController extends Controller
{
    use ChecksLogin;

    private const SINGULAR = 'account';
    private const ROUTE_SINGULAR = 'bank-' . self::SINGULAR;

    public function index(Request $request): View|RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;

            try {
                $this->_authorize($request, 'view bank account');

                $t = microtime(true);
                $accounts = BankAccount::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())->get();
                $this->logExecutionTime($t, $action, 'loadAccounts');

                $view = self::ROUTE_SINGULAR . '.' . $func;
                if (!ViewFacade::exists($view)) {
                    return defaultUndefinedException($request, new \Exception('view'), $action, route(self::ROUTE_SINGULAR . '.index'));
                }

                return ViewFacade::make($view, ['accounts' => $accounts]);
            } catch (AuthorizationException $e) {
                return defaultPermissionDenial($request, $e, $action);
            } catch (\Throwable $e) {
                Log::error($action . ' error', ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $action);
            }
        }, [UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
    }

    public function create(Request $request): View|RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            try {
                $this->_authorize($request, 'create bank account');
                $creatorId = $request->user()->creatorId();

                $t = microtime(true);
                $chartAccounts = ChartOfAccount::selectRaw('CONCAT(code," - ",name) AS code_name,id')
                    ->where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)
                    ->pluck('code_name', 'id')
                    ->prepend(__('Select Account'), '');
                $customFields = CustomField::where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)
                    ->where('module', self::SINGULAR)
                    ->get();
                $this->logExecutionTime($t, $action, 'loadCreateFormData');

                $view = self::ROUTE_SINGULAR . '.' . $func;
                if (!ViewFacade::exists($view)) {
                    return defaultUndefinedException($request, new \Exception('view'), $action, route(self::ROUTE_SINGULAR . '.index'));
                }

                return ViewFacade::make($view, compact('chartAccounts', 'customFields'));
            } catch (AuthorizationException $e) {
                return defaultPermissionDenial($request, $e, $action);
            } catch (\Throwable $e) {
                Log::error($action . ' error', ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $action);
            }
        }, [UsersConstants::COL_USER_ID => $request->user()?->id ?? null]);
    }

    public function store(Request $request): RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            try {
                $this->_authorize($request, 'create bank account');
                $creatorId = $request->user()->creatorId();

                $v = Validator::make($request->all(), [
                    BanksConstants::COL_HNM   => 'required|string',
                    BanksConstants::COL_NM    => 'required|string',
                    // unique:table,column,NULL,id,created_by,{$creatorId}
                    BanksConstants::COL_ACC_N => 'required|string|unique:' .
                        DatabaseConstants::TABLE_BANK_ACC . ',' . BanksConstants::COL_ACC_N .
                        ',NULL,id,' . DatabaseConstants::COL_TABLE_CREATOR . ',' . $creatorId,
                    BanksConstants::COL_OB    => 'required|numeric',
                    BanksConstants::COL_CT    => 'required|regex:/^([0-9\s\-\+\(\)]*)$/',
                ]);
                if ($v->fails()) {
                    return redirect()->route(self::ROUTE_SINGULAR . '.index')
                        ->with('error', $v->errors()->first());
                }

                DB::beginTransaction();
                $t = microtime(true);
                $data = $request->only([
                    BanksConstants::COL_COA,
                    BanksConstants::COL_HNM,
                    BanksConstants::COL_NM,
                    BanksConstants::COL_ACC_N,
                    BanksConstants::COL_OB,
                    BanksConstants::COL_CT,
                    BanksConstants::COL_ADR
                ]) + [DatabaseConstants::COL_TABLE_CREATOR => $creatorId];

                $account = BankAccount::create($data);
                CustomField::saveData($account, $request->customField);

                $this->logExecutionTime($t, $action, 'createAccount');
                DB::commit();

                return redirect()->route(self::ROUTE_SINGULAR . '.index')
                    ->with('success', __('Account successfully created.'));
            } catch (AuthorizationException $e) {
                DB::rollBack();
                return defaultPermissionDenial($request, $e, $action);
            } catch (\Throwable $e) {
                DB::rollBack();
                Log::error($action . ' failed', ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $action);
            }
        }, ['input' => $request->only(BanksConstants::COL_NM, BanksConstants::COL_ACC_N)]);
    }

    public function show(Request $request, BankAccount $bankAccount): RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $bankAccount, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            try {
                $this->_authorize($request, 'view bank account');
                if ($bankAccount->created_by !== $request->user()->creatorId()) {
                    throw new AuthorizationException;
                }
                return redirect()->route(self::ROUTE_SINGULAR . '.index');
            } catch (AuthorizationException $e) {
                return defaultPermissionDenial($request, $e, $action);
            } catch (\Throwable $e) {
                Log::error($action . ' failed', ['error' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $action);
            }
        }, ['bank_account_id' => $bankAccount->id ?? null]);
    }

    public function edit(Request $request, BankAccount $bankAccount): View|RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $bankAccount, $func, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            try {
                $this->_authorize($request, 'edit bank account');
                if ($bankAccount->created_by !== $request->user()->creatorId()) {
                    throw new AuthorizationException;
                }

                $creatorId = $request->user()->creatorId();

                $t = microtime(true);
                $chartAccounts = ChartOfAccount::selectRaw('CONCAT(code," - ",name) AS code_name,id')
                    ->where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)
                    ->pluck('code_name', 'id')
                    ->prepend(__('Select Account'), '');
                $bankAccount->customField = CustomField::getData($bankAccount, self::SINGULAR);
                $customFields = CustomField::where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)
                    ->where('module', self::SINGULAR)
                    ->get();
                $this->logExecutionTime($t, $action, 'loadEditFormData');

                $view = self::ROUTE_SINGULAR . '.' . $func;
                if (!ViewFacade::exists($view)) {
                    return defaultUndefinedException($request, new \Exception('view'), $action, route(self::ROUTE_SINGULAR . '.index'));
                }

                return ViewFacade::make($view, compact('bankAccount', 'chartAccounts', 'customFields'));
            } catch (AuthorizationException $e) {
                return defaultPermissionDenial($request, $e, $action);
            } catch (\Throwable $e) {
                Log::error($action . ' error', ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $action);
            }
        }, ['bank_account_id' => $bankAccount->id ?? null]);
    }

    public function update(Request $request, BankAccount $bankAccount): RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $bankAccount, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            try {
                $this->_authorize($request, 'edit bank account');
                $creatorId = $request->user()->creatorId();

                $lock = BankAccount::where('id', $bankAccount->id)->lockForUpdate()->firstOrFail();

                $v = Validator::make($request->all(), [
                    BanksConstants::COL_HNM   => 'required|string',
                    BanksConstants::COL_NM    => 'required|string',
                    // unique:table,column,{ignore_id},id,created_by,{$creatorId}
                    BanksConstants::COL_ACC_N => 'required|string|unique:' .
                        DatabaseConstants::TABLE_BANK_ACC . ',' . BanksConstants::COL_ACC_N . ',' .
                        $bankAccount->id . ',id,' . DatabaseConstants::COL_TABLE_CREATOR . ',' . $creatorId,
                    BanksConstants::COL_OB    => 'required|numeric',
                    BanksConstants::COL_CT    => 'required|regex:/^([0-9\s\-\+\(\)]*)$/',
                ]);
                if ($v->fails()) {
                    return redirect()->route(self::ROUTE_SINGULAR . '.index')
                        ->with('error', $v->errors()->first());
                }

                DB::beginTransaction();
                $t = microtime(true);
                $data = $request->only([
                    BanksConstants::COL_COA,
                    BanksConstants::COL_HNM,
                    BanksConstants::COL_NM,
                    BanksConstants::COL_ACC_N,
                    BanksConstants::COL_OB,
                    BanksConstants::COL_CT,
                    BanksConstants::COL_ADR
                ]) + [DatabaseConstants::COL_TABLE_CREATOR => $creatorId];

                $bankAccount->update($data);
                CustomField::saveData($bankAccount, $request->customField);

                $this->logExecutionTime($t, $action, 'updateAccount');
                DB::commit();

                return redirect()->route(self::ROUTE_SINGULAR . '.index')
                    ->with('success', __('Account successfully updated.'));
            } catch (AuthorizationException $e) {
                DB::rollBack();
                return defaultPermissionDenial($request, $e, $action);
            } catch (\Throwable $e) {
                DB::rollBack();
                Log::error($action . ' failed', ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $action);
            }
        }, ['bank_account_id' => $bankAccount->id ?? null, 'input' => $request->only(BanksConstants::COL_NM, BanksConstants::COL_ACC_N)]);
    }

    public function destroy(Request $request, BankAccount $bankAccount): RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($request, $bankAccount, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            try {
                $this->_authorize($request, 'delete bank account');

                DB::beginTransaction();
                $t = microtime(true);

                $lock = BankAccount::where('id', $bankAccount->id)->lockForUpdate()->firstOrFail();
                if ($lock->created_by !== $request->user()->creatorId()) {
                    throw new AuthorizationException;
                }

                $relationsExist = (
                    Revenue::where(BanksConstants::COL_REL_ID, $lock->id)->exists()
                    || InvoicePayment::where(BanksConstants::COL_REL_ID, $lock->id)->exists()
                    || Transaction::where(self::SINGULAR, $lock->id)->exists()
                    || Payment::where(BanksConstants::COL_REL_ID, $lock->id)->exists()
                    || BillPayment::where(BanksConstants::COL_REL_ID, $lock->id)->exists()
                );

                if ($relationsExist) {
                    DB::rollBack();
                    return redirect()->route(self::ROUTE_SINGULAR . '.index')
                        ->with('error', __('Please delete related records first.'));
                }

                $lock->delete();
                $this->logExecutionTime($t, $action, 'deleteAccount');

                DB::commit();
                return redirect()->route(self::ROUTE_SINGULAR . '.index')
                    ->with('success', __('Account successfully deleted.'));
            } catch (AuthorizationException $e) {
                DB::rollBack();
                return defaultPermissionDenial($request, $e, $action);
            } catch (\Throwable $e) {
                DB::rollBack();
                Log::error($action . ' failed', ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, $action);
            }
        }, ['bank_account_id' => $bankAccount->id ?? null]);
    }

    /**
     * Authorize the current user for a given ability.
     *
     * @throws AuthorizationException
     */
    private function _authorize(Request $request, string $ability): void
    {
        if (!$request->user()?->can($ability)) {
            Log::warning(__METHOD__ . ' permission denied', [
                'user_id' => Auth::id(),
                'ability' => $ability,
            ]);
            throw new AuthorizationException("Unauthorized: {$ability}");
        }
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/companies/bank_account_controller.py
import logging
from django.contrib import messages
from django.contrib.auth.decorators import login_required
from django.db import transaction
from django.forms import ModelForm
from django.utils.decorators import method_decorator
from django.shortcuts import get_object_or_404, redirect, render
from django.views.decorators.http import require_http_methods
from django.http import HttpRequest, HttpResponse
from .._helpers.http import get_redirect_url
from django.core.exceptions import PermissionDenied
from ....configs.messages_templates import get_exception_class_message

logger = logging.getLogger(__name__)

class BankAccountController(ModelForm):
  class Meta:
    model = None  # placeholder, set dynamically in __init__
  def __init__(self, *args: any, **kwargs: any) -> None:
    from ....Models.companies.bank_account import BankAccount
    self._meta.model = BankAccount
    super().__init__(*args, **kwargs)

@method_decorator(login_required)
def index(request: HttpRequest) -> HttpResponse:
  try:
    if not request.user.has_perm('create_bank_account'):
      messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
      return redirect(get_redirect_url(request))
    from ....Models.companies.bank_account import BankAccount
    accounts = BankAccount.objects.filter(created_by=request.user.creator_id)
    return render(request, 'bank_account/index.html', {'accounts': accounts})
  except Exception as e:
    logger.error("Failed in index: %s", e)
    messages.error(request, "An error occurred.")
    return redirect(get_redirect_url(request))

@method_decorator(login_required)
def create(request: HttpRequest) -> HttpResponse:
  try:
    if not request.user.has_perm('create_bank_account'):
      messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
      return redirect(get_redirect_url(request))
    from ....Models.charts.chart_of_account import ChartOfAccount
    from ....Models.shapes.custom_field import CustomField
    chart_accounts = ChartOfAccount.objects.filter(created_by=request.user.creator_id)
    chart_account_choices = [('', 'Select Account')] + [(acc.id, f"{acc.code} - {acc.name}") for acc in chart_accounts]
    custom_fields = CustomField.get_data('account', request.user.creator_id)
    return render(request, 'bank_account/create.html', {
      'chart_accounts': chart_account_choices,
      'custom_fields': custom_fields
    })
  except Exception as e:
    logger.error("Failed in create: %s", e)
    messages.error(request, "An error occurred.")
    return redirect(get_redirect_url(request))

@method_decorator(login_required)
@require_http_methods(["POST"])
def store(request: HttpRequest) -> HttpResponse:
  try:
    if not request.user.has_perm('create_bank_account'):
      messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
      return redirect(get_redirect_url(request))
    from ....Models.companies.bank_account import BankAccount
    from ....Models.shapes.custom_field import CustomField
    class _DynamicBankAccountForm(BankAccountController):
      class Meta(BankAccountController.Meta):
        model = BankAccount
    with transaction.atomic():
      form = _DynamicBankAccountForm(request.POST)
      if not form.is_valid():
        messages.error(request, next(iter(form.errors.values()))[0])
        return redirect('bank_account_index')
      account = form.save(commit=False)
      account.created_by = request.user.creator_id
      account.save()
      CustomField.save_data(account, request.POST.get('customField'))
    messages.success(request, "Account successfully created.")
    return redirect('bank_account_index')
  except Exception as e:
    logger.error("Failed in store: %s", e)
    messages.error(request, "An error occurred.")
    return redirect(get_redirect_url(request))

@method_decorator(login_required)
def edit(request: HttpRequest, id: int) -> HttpResponse:
  try:
    from ....Models.companies.bank_account import BankAccount
    from ....Models.charts.chart_of_account import ChartOfAccount
    from ....Models.shapes.custom_field import CustomField
    account = get_object_or_404(BankAccount, id=id)
    if not request.user.has_perm('edit_bank_account') or account.created_by != request.user.creator_id:
      messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
      return redirect(get_redirect_url(request))
    chart_accounts = ChartOfAccount.objects.filter(created_by=request.user.creator_id)
    chart_account_choices = [('', 'Select Account')] + [(acc.id, f"{acc.code} - {acc.name}") for acc in chart_accounts]
    custom_fields = CustomField.get_fields('account', request.user.creator_id)
    account.custom_fields = CustomField.get_data(account, 'account')
    return render(request, 'bank_account/edit.html', {
      'bank_account': account,
      'chart_accounts': chart_account_choices,
      'custom_fields': custom_fields
    })
  except Exception as e:
    logger.error("Failed in edit: %s", e)
    messages.error(request, "An error occurred.")
    return redirect(get_redirect_url(request))

@method_decorator(login_required)
@require_http_methods(["POST"])
def update(request: HttpRequest, id: int) -> HttpResponse:
  try:
    from ....Models.companies.bank_account import BankAccount
    from ....Models.shapes.custom_field import CustomField
    account = get_object_or_404(BankAccount, id=id)
    if not request.user.has_perm('edit_bank_account') or account.created_by != request.user.creator_id:
      messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
      return redirect(get_redirect_url(request))
    class _DynamicBankAccountForm(BankAccountController):
      class Meta(BankAccountController.Meta):
        model = BankAccount
    with transaction.atomic():
      form = _DynamicBankAccountForm(request.POST, instance=account)
      if not form.is_valid():
        messages.error(request, next(iter(form.errors.values()))[0])
        return redirect('bank_account_index')
      account = form.save(commit=False)
      account.created_by = request.user.creator_id
      account.save()
      CustomField.save_data(account, request.POST.get('customField'))
    messages.success(request, "Account successfully updated.")
    return redirect('bank_account_index')
  except Exception as e:
    logger.error("Failed in update: %s", e)
    messages.error(request, "An error occurred.")
    return redirect(get_redirect_url(request))

@method_decorator(login_required)
def destroy(request: HttpRequest, id: int) -> HttpResponse:
  try:
    from ....Models.companies.bank_account import BankAccount
    from ....Models.bills.revenue import Revenue
    from ....Models.bills.invoice_payment import InvoicePayment
    from ....Models.bills.payment import Payment
    from ....Models.bills.bill_payment import BillPayment
    from ....Models.bills.transaction import Transaction
    account = get_object_or_404(BankAccount, id=id)
    if not request.user.has_perm('delete_bank_account') or account.created_by != request.user.creator_id:
      messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
      return redirect(get_redirect_url(request))
    has_related = any([
      Revenue.objects.filter(account_id=account.id).exists(),
      InvoicePayment.objects.filter(account_id=account.id).exists(),
      Transaction.objects.filter(account_id=account.id).exists(),
      Payment.objects.filter(account_id=account.id).exists(),
      BillPayment.objects.exists()
    ])
    if has_related:
      messages.error(request, "Please delete related record of this account.")
      return redirect('bank_account_index')
    account.delete()
    messages.success(request, "Account successfully deleted.")
    return redirect('bank_account_index')
  except Exception as e:
    logger.error("Failed in destroy: %s", e)
    messages.error(request, "An error occurred.")
    return redirect(get_redirect_url(request))
